pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart

```diff
@@ -1,68 +1,141 @@
   const PokemonOverviewPage({
     required this.pokemons,
+    required this.searchedPokemons,
+    required this.filterPokemons,
+    required this.clearSearchedPokemons,
     Key? key,
   }) : super(key: key);
 
   final Async<List<Pokemon>> pokemons;
+  final List<Pokemon> searchedPokemons;
+  final VoidCallback clearSearchedPokemons;
+  final Function(String) filterPokemons;
+
+  @override
+  State<PokemonOverviewPage> createState() => _PokemonOverviewPageState();
+}
+
+class _PokemonOverviewPageState extends State<PokemonOverviewPage> {
+  late TextEditingController _searchController;
+  late bool isSearching;
+  Timer? _debounceTimer;
+  late String _searchQuery = emptyString;
+
+  @override
+  void initState() {
+    super.initState();
+    isSearching = false;
+    _searchController = TextEditingController();
+  }
+
+  @override
+  void dispose() {
+    _searchController.dispose();
+    _debounceTimer?.cancel();
+    if (widget.searchedPokemons.isNotEmpty) {
+      widget.clearSearchedPokemons();
+    }
+    super.dispose();
+  }
 
   @override
   Widget build(BuildContext context) {
     return Scaffold(
       appBar: const CustomAppBar(title: pokemonOverviewTitle),
-      body: pokemons.when(
-        (data) {
-          return GridView.builder(
-            itemCount: data.length,
-            gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(crossAxisCount: 2),
-            itemBuilder: (_, index) {
-              final pokemon = data[index];
-              return PokemonCard(
-                pokemon: pokemon,
-                onTap: () => _navigateToPokemonDetailsPage(context: context, pokemon: pokemon),
-              );
+      body: Column(
+        children: [
+          PokemonSearchBar(
+            textController: _searchController,
+            clearSearchBar: _clearSearchedPokemons,
+            textChanged: (val) {
+              setState(() {
+                _searchQuery = val;
+              });
+              _duringPokemonSearch(val);
             },
-          );
-        },
-        loading: () => Container(
-          color: mainBgColor,
-          child: Center(
-            child: Image.asset(
-              pikachuRunningImage,
-              height: 210,
-              width: 210,
-            ),
           ),
-        ),
-        error: (errorMessage) {
-          WidgetsBinding.instance.addPostFrameCallback(
-            (_) => _showErrorMessageSnackbar(context, errorMessage),
-          );
-          return Center(
-            child: Column(
-              mainAxisAlignment: MainAxisAlignment.center,
-              children: [
-                const Text(noPokemonsAvailableLabel),
-                const VerticalSpace(height: 20),
-                Image.asset(
-                  pokedexErrorImage,
-                  height: 120,
-                  width: 120,
+          Expanded(
+            child: widget.pokemons.when(
+              (data) {
+                final filteredPokemons =
```

Comment:
no need for this since this is already been handled in the vm together with your action 🙂